content/src/controller.js

```diff
@@ -786,6 +793,87 @@ function saveAction(forceOverwrite, loginPrompt, doneCallback) {
   });
 }
 
+function generateThumbnailDataUrl() {
+  var THUMBNAIL_SIZE = 128;
+  // Use the same background as the default icons.
+  var BACKGROUND_COLOR = '#eeeeee';
+
+  // Get the canvas inside the iframe.
+  var iframe = document.getElementsByTagName('iframe')[0];
+  var innerDoc = iframe.contentDocument || iframe.contentWindow.document;
+  var canvas = innerDoc.getElementsByTagName('canvas')[0];
+  var w = canvas.width;
+  var h = canvas.height;
+
+  if (canvas.id === 'turtle') {
+    return;
+  }
+
+  // Get the image data.
```

Comment:
The next 40 lines or so (cropping and resizing a canvas to a specific dimension) should probably be its own function.